tests/integration/models/device.spec.coffee

```diff
@@ -1369,3 +1450,235 @@ describe 'Device Model', ->
+					mockDevice = {
+						uuid
+						device_type: 'raspberrypi3'
+						is_online: true
+						os_version
+						os_variant
+					}
+
+					m.chai.expect(-> balena.models.device._checkOsUpdateTarget(
+						mockDevice
+						'1.25.0'
+					)).to.throw('Target OS version must be >= 1.26.0')
+
+			it 'should not throw when it is a valid v1 -> v1 hup', ->
+				[
+					['Resin OS 1.8.0', '']
+					['Resin OS 1.10.0', '']
+					['Resin OS 1.19.0', '']
+					['Resin OS 1.21.0', '']
+				].forEach ([os_version, os_variant]) ->
+					mockDevice = {
+						uuid
+						device_type: 'raspberrypi3'
+						is_online: true
+						os_version
+						os_variant
+					}
+
+					m.chai.expect(-> balena.models.device._checkOsUpdateTarget(
+						mockDevice
+						'1.28.0'
+					)).to.not.throw()
+
+			it 'should not throw when it is a valid v1 -> v2 hup', ->
+				[
+					['Resin OS 1.8.0', '']
+					['Resin OS 1.10.0', '']
+					['Resin OS 1.19.0', '']
+					['Resin OS 1.21.0', '']
+				].forEach ([os_version, os_variant]) ->
+					mockDevice = {
+						uuid
+						device_type: 'raspberrypi3'
+						is_online: true
+						os_version
+						os_variant
+					}
+
+					m.chai.expect(-> balena.models.device._checkOsUpdateTarget(
+						mockDevice
+						'2.5.0+rev1'
+					)).to.not.throw()
+
+			it 'should throw when the device is offline', ->
+				[
+					['balenaOS 2.26.0+rev1', 'prod']
+				].forEach ([os_version, os_variant]) ->
+					mockDevice = {
+						uuid
+						device_type: 'raspberrypi3'
+						is_online: false
+						os_version
+						os_variant
+					}
+
+					m.chai.expect(-> balena.models.device._checkOsUpdateTarget(
+						mockDevice
+						'2.29.2+rev1.prod'
+					)).to.throw('The device is offline')
+
+			it 'should throw when the device is running a dev version', ->
+				[
+					['Resin OS 2.0.0+rev3 (dev)', 'dev']
+					['Resin OS 2.0.0+rev3 (dev)', '']
+					['Resin OS 2.0.0+rev5 (dev)', '']
+					['Resin OS 2.0.2+rev2', 'dev']
+					['Resin OS 2.0.3+rev1', 'dev']
+					['Resin OS 2.0.4+rev1', 'dev']
+					['Resin OS 2.0.5', 'dev']
+					['Resin OS 2.0.5+rev1', 'dev']
+					['Resin OS 2.0.6+rev1', 'dev']
+					['Resin OS 2.0.6+rev2', 'dev']
+					['Resin OS 2.1.0+rev1', 'dev']
+					['Resin OS 2.2.0+rev1', 'dev']
+					['Resin OS 2.9.7+rev1', 'dev']
+					['Resin OS 2.12.0+rev1', 'dev']
+					['Resin OS 2.12.1+rev1', 'dev']
+					['Resin OS 2.12.3', 'dev']
+					['Resin OS 2.12.3+rev1', 'dev']
+					['balenaOS 2.26.0', 'dev']
+					['balenaOS 2.26.0+rev1', 'dev']
+					['balenaOS 2.28.0+rev1', 'dev']
+				].forEach ([os_version, os_variant]) ->
+					mockDevice = {
+						uuid
+						device_type: 'raspberrypi3'
+						is_online: true
+						os_version
+						os_variant
+					}
+
+					m.chai.expect(-> balena.models.device._checkOsUpdateTarget(
+						mockDevice
+						'2.29.2+rev1.prod'
+					)).to.throw('Updates cannot be performed on development balenaOS variants')
+
+			it 'should throw when the device is running a pre-release version', ->
```

Comment:
I don't think it should throw an error, these versions should be updatable, IMHO.

Reply:
The proxy & the hup utils actually prohibit it, so I'm just replicating the same check in here, to make sure whether an update changes the behavior.
If that's indeed something supported, then we should first update it in the proxy.
See: https://github.com/balena-io/resin-proxy/blob/5efa5a905866fef31086c195e9d9ec4642eb1ee1/src/common/version-utils.ts#L66-L70
See: https://github.com/balena-io-modules/balena-hup-action-utils/blob/5032cf146ae8171781305881f4beb2ef1ef4ce8c/lib/index.ts#L80-L87